# VCF parse
Write functions to parse VCF from IonTorrent/Illumina for quality metrics of variants.
At first, these metrics will be used to assist variant filtering in an data-informed manner.
* Advantages
- more standardised outputs
- different software produce VCFs
* Disadvantages
- will need to create output usable for the diagnostic testing team.


# PROBLEM
not all VCFs from IR contain gene information. Will require mapping of genes but basically updated VCF annotation to implement VCF based gnomad feature etc.


# VCFparse
